src/smt/model_blocker.cpp

```diff
@@ -0,0 +1,292 @@
+/*********************                                                        */
+/*! \file model_blocker.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implementation of utility for blocking models.
+ **
+ **/
+
+#include "smt/model_blocker.h"
+
+#include "expr/node.h"
+#include "expr/node_algorithm.h"
+#include "theory/quantifiers/term_util.h"
+
+using namespace CVC4::kind;
+
+namespace CVC4 {
+
+Expr ModelBlocker::getModelBlocker(const std::vector<Expr>& assertions,
+                                   theory::TheoryModel* m,
+                                   BlockModelsMode mode,
+                                   const std::vector<Expr>& exprToBlock)
+{
+  NodeManager* nm = NodeManager::currentNM();
+  // convert to nodes
+  std::vector<Node> tlAsserts;
+  for (unsigned i = 0, size = assertions.size(); i < size; i++)
+  {
+    Node a = Node::fromExpr(assertions[i]);
+    tlAsserts.push_back(a);
+  }
+  std::vector<Node> nodesToBlock;
+  for (unsigned i = 0, size = exprToBlock.size(); i < size; i++)
+  {
+    nodesToBlock.push_back(Node::fromExpr(exprToBlock[i]));
+  }
+  Trace("model-blocker") << "Compute model blocker, assertions:" << std::endl;
+  Node blocker;
+  if (mode == BLOCK_MODELS_LITERALS)
+  {
+    Assert(nodesToBlock.empty());
+    // optimization: filter to only top-level disjunctions
```

Comment:
What about implications? Why aren't they needed?

Reply:
Implications are disjunctions, I'll clarify the comment.